src/storage/v2/disk/storage.cpp

```diff
@@ -540,61 +540,110 @@ VerticesIterable DiskStorage::DiskAccessor::Vertices(View view) {
 
 VerticesIterable DiskStorage::DiskAccessor::Vertices(LabelId label, View view) {
   auto *disk_storage = static_cast<DiskStorage *>(storage_);
+  const auto cache_key = label;
 
   if (disk_storage->edge_import_status_ == EdgeImportMode::ACTIVE) {
     HandleLoadingLabelForEdgeImportCache(label);
 
     return VerticesIterable(disk_storage->edge_import_mode_cache_->Vertices(label, view, storage_, &transaction_));
   }
 
-  transaction_.index_storage_.emplace_back(std::make_unique<utils::SkipList<storage::Vertex>>());
-  auto &indexed_vertices = transaction_.index_storage_.back();
-  transaction_.index_deltas_storage_.emplace_back();
-  auto &index_deltas = transaction_.index_deltas_storage_.back();
+  utils::SkipList<memgraph::storage::Vertex> *indexed_vertices{nullptr};
+  std::list<storage::Delta> *index_deltas{nullptr};
 
-  auto gids = MergeVerticesFromMainCacheWithLabelIndexCache(label, view, index_deltas, indexed_vertices.get());
-  LoadVerticesFromDiskLabelIndex(label, gids, index_deltas, indexed_vertices.get());
+  auto merge_with_main_cache = [&](auto &index, auto &index_delta_storage) -> std::unordered_set<Gid> {
+    index[cache_key] = utils::SkipList<Vertex>();
+    indexed_vertices = &index[cache_key];
+    index_delta_storage.emplace_back();
+    index_deltas = &index_delta_storage.back();
+
+    return MergeVerticesFromMainCacheWithLabelIndexCache(cache_key, view, *index_deltas, *indexed_vertices);
+  };
+
+  auto &label_index_cache = transaction_.label_index_cache_;
+  auto &label_index_cache_ci = transaction_.label_index_cache_ci_;
+  auto &index_delta_storage = transaction_.index_deltas_storage_;
+
+  if (transaction_.command_id > label_index_cache_ci) {
+    label_index_cache_ci = transaction_.command_id;
+
+    auto gids = merge_with_main_cache(label_index_cache, index_delta_storage);
+    LoadVerticesFromDiskLabelIndex(gids, cache_key, *index_deltas, *indexed_vertices);
+  } else {
+    if (label_index_cache.contains(cache_key) && view == View::OLD) {
+      merge_with_main_cache(label_index_cache, index_delta_storage);
+    } else {
+      auto gids = merge_with_main_cache(label_index_cache, index_delta_storage);
+      LoadVerticesFromDiskLabelIndex(gids, cache_key, *index_deltas, *indexed_vertices);
+    }
+  }
```

Comment:
I don't think this is correct logic but may be wrong. First I would suggest removing these 3 variables:
```
auto &label_index_cache = transaction_.label_index_cache_;
auto &label_index_cache_ci = transaction_.label_index_cache_ci_;
auto &index_delta_storage = transaction_.index_deltas_storage_;
```
I think it just makes code harder to read. Whether you write in the code transaction_.label_index_cache_ or label_index_cache no big difference.
Second thing, `merge_with_main_cache` can be extracted since you can call it in all if-else branches. 
Third thing and the most important:
I don't see at any place that you return cached vertices instead of going to disk. If condition
`if (label_index_cache.contains(cache_key) && view == View::OLD)` is true, you call just `merge_with_main_cache` but in that method you already allocated new SkipList and new deltas, you didn't reuse anything. So not sure how currently this PR improves logic.

Reply:
This comment applies for all other methods too

Reply:
1. It is not much more to write, but I do not agree with you that it is not easier to read. I changed the reference names to be more concise, please check it and let me know what you think.
2. `merge_with_main_cache` is not that convenient to extract, since the 4 index-overloaded `Vertices(...)` functions are using different function calls to do the merge itself, where the functions are members of the `Accessor` class, meaning they are not convenient to be abstracted into functors(you will have to `std::bind` them or wrap them into a lambda to be able to pass them because of `this`). Using std::function to type erase might be a solution but the signature of the used member functions are different, so it would be harder to read it either way when providing a generic abstraction in my opinion. Since this implementation is likely to change once we introduce the proper on-disk indexing mechanism, I would rather keep it in this form. 
3. We already discussed this so I am leaving this here as information for other people. The caching mechanism is changing in a way where data from disk is read into the main cache which is bound to the lifetime of any given transaction and the index caches are merged with the main cache where appropriate.